### PR TITLE
Add support for TMAG5170 3d Hall effect sensor

### PR DESCRIPTION
**Pull request by morsisko**
Hello,

this pull request adds support for Texas Instruments' TMAG5170 high-precision, linear 3D Hall-effect sensor with SPI bus interface.
For now following features are implemented:
- Reading 3 different magnetic channels
- Reading temperature
- Reading angular rotation of magnet
- Data ready triggers
- Various operating modes (continuous conversion, manual trigger)
- Power management (deep sleep mode)
- Error detection in TX/RX data using CRC
Most of the configuration is done via devicetree and not Kconfig, this is because in my opinion this gives the end user more flexibility, as they can connect multiple TMAGs with different settings (for example one sensor with fast sampling, another one for slow and redundancy)

Also I'm not sure about the CRC. As for now Zephyr is missing CRC4, I've added two CRC4 algorithms. One that is specific to TMAG (crc4_ti) and another one that is generic. They would both work for this sensor, one is just faster. Not sure if I should leave the `crc4_ti`, or stick to `crc4`. Maybe I shouldn't touch the lib/os/crc library, instead implement the crc algorithm directly in the sensor driver code? I'm open for suggestions.

### CODE REVIEW

**Comment by morsisko** [comment]
I would be thankful for a hint on how to resolve the unit tests issues. Did I forgot to add the sensor somewhere? The device tree error is unclear to me

File: tests/drivers/build_all/sensor/spi.dtsi [hunk 1]
```diff
@@ -285,3 +285,11 @@ test_spi_bmi08x_gyro: bmi08x@24 {
 	gyro-hz = "1000_116";
 	gyro-fs = <1000>;
 };
+
+test_spi_tmag5170: tmag5170@25 {
+	compatible = "ti,tmag5170";
+	reg = <0x25>;
```
Comment (henrikbrixandersen) [inline tests/drivers/build_all/sensor/spi.dtsi:291]: You need to add an extra line to `cs-gpios` in `tests/drivers/build_all/sensor/app.overlay` for the added CS GPIO.

Comment (morsisko) [reply tests/drivers/build_all/sensor/spi.dtsi:291]: Thanks, solved

**Review by MaureenHelm — DISMISSED** [review]
Looks great, thank you. Please rebase and resolve the merge conflict

**Comment by morsisko** [comment]
Conflicts has been resolved, PR is now rebased

**Review by MaureenHelm — APPROVED** [review]

**Review by vaishnavachath — APPROVED** [review]

**Pull request merged**